client.go

```diff
@@ -247,69 +223,264 @@ func (clt *Client) userPassAuthentication(proxyConn net.Conn) (err error) {
+	}
+	if tmp[1] != Granted {
+		return nil, errors.New("server refuse client request")
+	}
+	// Read address
+	replyAddr, _, err := readAddress(proxyConn, request.VER)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get reply address, %v", err)
+	}
+	return replyAddr, nil
+}
+
+// UDPForward socks UDP forward, get a udp connect
+func (clt *Client) UDPForward(request *Request) *net.UDPConn {
+	if request.CMD != UDP_ASSOCIATE {
+		clt.logf()("command should be udp_associate")
+		return nil
+	}
+	proxyTCPConn, udpserverBoundAddr := clt.TCPConnect(request)
+	if proxyTCPConn == nil {
+		clt.logf()("TCP conn failure")
+		return nil
+	}
+	// TCP conn are no longer needed,close
+	defer proxyTCPConn.Close()
+	serverBoundUDPAddr, err := net.ResolveUDPAddr("udp", udpserverBoundAddr.String())
+	if err != nil {
+		clt.logf()(err.Error())
+		return nil
+	}
+	// Get local UDP addr
+	localTCPAddr := proxyTCPConn.LocalAddr().(*net.TCPAddr)
+	localUDPAddr := &net.UDPAddr{
+		IP:   request.Addr,
+		Port: int(request.Port),
+		Zone: localTCPAddr.Zone,
+	}
+	// Get UDP Conn
+	proxyUDPConn, err := net.DialUDP("udp", localUDPAddr, serverBoundUDPAddr)
+	if err != nil {
+		clt.logf()(err.Error())
+		return nil
+	}
+	if clt.TimeOut != 0 {
+		err = proxyUDPConn.SetDeadline(time.Now().Add(clt.TimeOut))
 		if err != nil {
-			return proxyConn, err
+			clt.logf()(err.Error())
 		}
-		// The client connects to the SOCKS server and sends a CONNECT request when it wants to establish a connection to an application server.
-		// The client includes in the request packet the IP address and the port number of the destination host, and userid, in the following format.
-		//    +----+----+----+----+----+----+----+----+----+----+....+----+
-		//    | VN | CD | DSTPORT |      DSTIP        | USERID       |NULL|
-		//    +----+----+----+----+----+----+----+----+----+----+....+----+
-		//      1    1      2              4           variable       1
-		if _, err := proxyConn.Write(append([]byte{request.VER, request.CMD}, destAddrByte...)); err != nil {
-			return proxyConn, err
-		}
-		// A reply packet is sent to the client when this connection is established,or when the request is rejected or the operation fails.
-		//    +----+----+----+----+----+----+----+----+
-		//    | VN | CD | DSTPORT |      DSTIP        |
-		//    +----+----+----+----+----+----+----+----+
-		//       1    1      2              4
-		tmp, err := ReadNBytes(proxyConn, 2)
+	}
+	return proxyUDPConn
+}
+
+//Bind socks bind cmd, get socks serve Listening bind addr,second time reply err,bind connect
+func (clt *Client) Bind(ver VER, destAddr *Address) (*Address, <-chan error, net.Conn) {
+	request := &Request{
+		Address: destAddr,
+		CMD:     BIND,
+		VER:     ver,
+	}
+	proxyBindConn, err := net.Dial("tcp", clt.ProxyAddr)
+	if err != nil {
+		clt.logf()(err.Error())
+		return nil, nil, nil
+	}
+	if clt.TimeOut != 0 {
+		err = proxyBindConn.SetDeadline(time.Now().Add(clt.TimeOut))
```

Comment:
After **handshake**, you should use net.Conn.SetReadline(time.Time{}) to cancel conn's timeout limit.